---
title: "DEG"
author: "Marco Tello"
date: "2024-08-21"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(data.table)
library(ggplot2)
library(ComplexHeatmap)
library(stringr)
library(ggrepel)

## Custom functions

custom_volcano <- function(DEG_results, alpha, log2FC, top_exp){
  # DESeq2 will make some adjusted pvalues NA, details in their docs
  num_na <- sum(is.na(DEG_results$padj))
  # Ommit na results and categorize each gene
  temp <- data.table(na.omit(DEG_results))
  temp[, alpha := ifelse(padj < alpha, TRUE, FALSE)]
  temp[, log2FC := ifelse(abs(log2FoldChange) >= log2FC, TRUE, FALSE)]
  temp[, DEG := "No"]
  temp[, DEG := ifelse(alpha & !log2FC, "FDR", DEG)]
  temp[, DEG := ifelse(alpha & log2FC, "FDR and FC", DEG)]
  # Include gene names for the highest fold changes
  # Genes not selected will have their name set to "" in the plot
  setorder(temp, "log2FoldChange")
  top_exp_genes <- temp[DEG == "FDR and FC", external_gene_name]
  top_exp_genes <- top_exp_genes[c(1:top_exp, (length(top_exp_genes) - top_exp -1):length(top_exp_genes))]
  temp[, gene_label := ifelse(external_gene_name %in% top_exp_genes, external_gene_name, "")]
  # Make plot
  ggplot(temp, aes(x = log2FoldChange, 
                   y = -log10(padj), 
                   color = DEG)) +
    geom_hline(yintercept = -log10(alpha), 
               linetype = "dashed", color = "gray", linewidth = 1) +
    geom_vline(xintercept = log2FC, 
               linetype = "dashed", color = "gray", linewidth = 1) +
    geom_vline(xintercept = -log2FC, 
               linetype = "dashed", color = "gray", linewidth = 1) +
    geom_point(alpha = 0.4, size = 3) +
    scale_x_continuous(breaks = seq(ceiling(range(temp$log2FoldChange))[1],
                                    ceiling(range(temp$log2FoldChange))[2])) +
    scale_color_manual(values = c("No" = "darkgray", 
                                  "FDR" = "blue", 
                                  "FDR and FC" = "red"), 
                       labels = c(paste("FDR (", 
                                        as.character(table(temp$DEG)[1]), 
                                        ")", sep = ""), 
                                  paste("FDR and FC (",
                                        as.character(table(temp$DEG)[2]),")", sep = ""),
                                  paste("Not Sig. (", 
                                        as.character(table(temp$DEG)[3] + num_na), ")", sep = ""))) +
    ylab(expression(-log[10]("FDR")))+
    xlab(expression(log[2]("Fold Change")))+
    geom_label_repel(aes(label = gene_label),
                     force = 20, nudge_x = 5, 
                      nudge_y = -20, colour = "black")+
    labs(color = "DEG status") +
    theme_bw()
}


visualize_degs <- function(DEG_results, alpha, log2FC, top_exp = 5){
  histogram <- ggplot(na.omit(DEG_results), aes(x = pvalue)) + 
    geom_histogram(binwidth = 0.05, breaks = seq(0, 1, by = 0.05), 
                   fill = "gray", color = "black") + 
    xlab(expression(italic("p-value")))+
    labs(title = "Histogram of p-values", y = "Frequency")+
    theme_bw()
  volcano <- custom_volcano(DEG_results, alpha, log2FC, top_exp)
  return(list(histogram, volcano))
}

```

# DEG visualization and comparison

We load the results from the differential expression analysis using DESeq2. The file contains 7 columns:

ENSG: ENSEMBL Gene Identifier
baseMean: Mean expression across samples
log2FoldChange: Change in expression in the experimental condition vs control condition
lfcSE:
stat: 
pvalue: Probability of observing a change in expression as extreme as "stat" assuming there were no changes between conditions
padj: Pvalue adjusted for multiple testing using FDR
Group: Comparison group label [Experimental condition]_Vs[Control condition]. Possible values: "H2O2_Vs_Ctrl" "GEN9_vs_H2O2" "PNS2_vs_H2O2" "PTS3_vs_H2O2"


```{r}
degs <- fread("~/GitHub/RNAseq_PECs/data/DEG_results.tsv")
exp_data <- fread("~/GitHub/RNAseq_PECs/data/filtered_counts.tsv")
exp_data <- merge.data.table(x = degs[,.SD, .SDcols = c("ensembl_gene_id_version", 
                                                        "external_gene_name",
                                                        "entrezgene_id")], 
                             y = exp_data, 
                             by.x = "ensembl_gene_id_version", 
                             by.y = "ENSG", 
                             all.y = TRUE)
exp_data <- unique(exp_data)

vst_counts <- DESeq2::varianceStabilizingTransformation(as.matrix(exp_data[, 4:ncol(exp_data)]))
vst_counts <- cbind(exp_data[,1:3], data.table(vst_counts))

head(degs)
head(vst_counts)

```

## Individual differential expression profiles.

We will first visualize DEGs by group using volcano plots. We will define DEGs as follow:

```{r}
alpha <- 0.01
log2FC <- 2
  
```



### H2O2_Vs_Ctrl

First we will visualize the changes in expression comparing exposure to H2O2 vs control condition.

We will first generate a histogram of the raw p-values for all genes. If the H0 is true, we would expect to see an uniform distribution of pvalues. A peak at low zeroes indicate we reject the H0.

```{r}
temp <- visualize_degs(degs[Group == "H2O2_Vs_Ctrl"], alpha, log2FC, top_exp = 5)
temp[[1]]
```
Then we visualize the expression changes using a volcano plot.

```{r}
temp[[2]]
```


### GEN9_vs_H2O2

First we will visualize the changes in expression comparing exposure to H2O2 vs control condition.

We will first generate a histogram of the raw p-values for all genes. If the H0 is true, we would expect to see an uniform distribution of pvalues. A peak at low zeroes indicate we reject the H0.

```{r}
temp <- visualize_degs(degs[Group == "GEN9_vs_H2O2"], alpha, log2FC)
temp[[1]]
```
Then we visualize the expression changes using a volcano plot.

```{r}
temp[[2]]
```


### PNS2_vs_H2O2

First we will visualize the changes in expression comparing exposure to H2O2 vs control condition.

We will first generate a histogram of the raw p-values for all genes. If the H0 is true, we would expect to see an uniform distribution of pvalues. A peak at low zeroes indicate we reject the H0.

```{r}
temp <- visualize_degs(degs[Group == "PNS2_vs_H2O2"], alpha, log2FC)
temp[[1]]
```
Then we visualize the expression changes using a volcano plot.

```{r}
temp[[2]]
```


### PTS3_vs_H2O2

First we will visualize the changes in expression comparing exposure to H2O2 vs control condition.

We will first generate a histogram of the raw p-values for all genes. If the H0 is true, we would expect to see an uniform distribution of pvalues. A peak at low zeroes indicate we reject the H0.

```{r}
temp <- visualize_degs(degs[Group == "PTS3_vs_H2O2"], alpha, log2FC)
temp[[1]]
```
Then we visualize the expression changes using a volcano plot.

```{r}
temp[[2]]
```

### TODO: Explore samples in the conditions PTS3 and H2O2. Are ther a lot of genes with zero reads that scape filtering step?



## Visualize shared set of DEGs

Based on our final selection of FDR and Fold Change cutoffs we will determine what genes are shared across conditions using an upset plot. 

```{r}
# Select for DEGs based on alpha and log2FC
degs_sig <- degs[padj < alpha & abs(log2FoldChange) >= log2FC]
# Reshape data.table
gene_mat <- dcast(data = degs_sig[, .(value = 1), 
                                  by = .(ensembl_gene_id_version, Group)], 
                formula = ensembl_gene_id_version ~ Group, 
                value.var = "value", 
                fill = 0)
# Convert to matrix for compatibility with UpSet
gene_mat <- as.matrix(gene_mat[, .SD, .SDcols = !"ensembl_gene_id_version"], rownames = gene_mat$ensembl_gene_id_version)

# Possible modes: 
# "distinct" (default), "intersect", "union"
comb_mat <- make_comb_mat(gene_mat, mode = "distinct")
UpSet(comb_mat)
```

Based on the UpSet plot, we can determine the genes that potentially could have reversed their expression change. 

```{r}
# Binary order:
# GEN9, H2O2, PNS2, PTS3
candidate_genes <- extract_comb(comb_mat, "1110")
for(combination in c("1101","0111","1100","0110")){
  candidate_genes <- unique(c(candidate_genes, extract_comb(comb_mat, combination)))
}
candidate_genes
```

